### PR TITLE
break: Revert "chore: Revert "fix: upload-artifact and download-artifact v4""

### PR DESCRIPTION
**Pull request by ramonpetgrave64**
Reverts slsa-framework/slsa-github-generator#3398

Following up with 

- https://github.com/slsa-framework/example-trw/pull/3
- https://github.com/slsa-framework/example-package/pull/340
- https://github.com/slsa-framework/slsa-verifier/pull/719

### CODE REVIEW

**Review by laurentsimon — APPROVED** [review]
Thanks. I leave the privilege of merging to you this time ;)

File: .github/workflows/pre-submit.e2e.generic.default.yml [hunk 1]
```diff
@@ -36,6 +36,7 @@ jobs:
       # echo "2e0390eb024a52963db7b95e84a9c2b12c004054a7bad9a97ec0c7c89d4681d2    binary-name" | base64 -w0
       base64-subjects: "MmUwMzkwZWIwMjRhNTI5NjNkYjdiOTVlODRhOWMyYjEyYzAwNDA1NGE3YmFkOWE5N2VjMGM3Yzg5ZDQ2ODFkMiAgICBiaW5hcnktbmFtZQo="
       compile-generator: true
+      provenance-name: build.intoto.jsonl
```
Comment (laurentsimon) [inline .github/workflows/pre-submit.e2e.generic.default.yml:39]: ah good catch! Were these effectively colluding with each other?

Comment (ramonpetgrave64) [reply .github/workflows/pre-submit.e2e.generic.default.yml:39]: Yes, which upload-artifact@v4 doesn't allow anymore.

Comment (ramonpetgrave64) [reply .github/workflows/pre-submit.e2e.generic.default.yml:39]: Before, I hadn't realized that this particular test was also failing.

**Pull request merged**